engine/access/rpc/backend/backend_network.go

```diff
@@ -80,6 +83,79 @@ func (b *backendNetwork) GetLatestProtocolStateSnapshot(_ context.Context) ([]by
 	return data, nil
 }
 
+// GetProtocolStateSnapshotByBlockID returns serializable Snapshot by blockID
+func (b *backendNetwork) GetProtocolStateSnapshotByBlockID(_ context.Context, blockID flow.Identifier) ([]byte, error) {
+	snapshotHeadByBlockId, err := b.state.AtBlockID(blockID).Head()
+	if err != nil {
+		if errors.Is(err, state.ErrUnknownSnapshotReference) {
+			return nil, status.Errorf(codes.NotFound, "failed to get a valid snapshot: block not found")
+		}
+
+		return nil, status.Errorf(codes.Internal, "could not get header by blockID: %v", err)
+	}
+
+	snapshotByHeight := b.state.AtHeight(snapshotHeadByBlockId.Height)
+	snapshotHeadByHeight, err := snapshotByHeight.Head()
+	if err != nil {
+		if errors.Is(err, state.ErrUnknownSnapshotReference) {
+			return nil, status.Errorf(codes.InvalidArgument,
+				"failed to retrieve snapshot for block by height %d: block not finalized", snapshotHeadByBlockId.Height)
+		}
```

Comment:
```suggestion
		if errors.Is(err, storage.ErrNotFound) {
			// The block exists, but no block has been finalized at its height. Therefore, this block
			// may be finalized in the future, and the client can retry.
			return nil, status.Errorf(codes. FailedPrecondition,
				"failed to retrieve snapshot for block by height %d: block not finalized and is above finalized height", snapshotHeadByBlockId.Height)
		}
```
Suggesting to use `FailedPrecondition` to differentiate the two "found block" failures cases.